src/main/java/com/comcast/zucchini/FlexibleBarrier.java

```diff
@@ -0,0 +1,234 @@
+package com.comcast.zucchini;
+
+import java.util.Collections;
+import java.util.Set;
+
+import java.lang.IllegalStateException;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.Phaser;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeoutException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is the heavy lifter behind the Barrier class.
+ *
+ * This integrates with the AbstractTestContext, the ZucchiniRuntime, and the TestContext to track when and what threads have failed, as well as kill the threads that should timeout.  The barrier will dynamically rezsize itself in accordance with the number of threads that can successfully finish.
+ *
+ * @author Andrew Benton
+ */
+class FlexibleBarrier {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FlexibleBarrier.class);
+
+    private AbstractZucchiniTest azt;
+    private Phaser primary;
+    private Phaser secondary;
+    private Set<TestContext> arrivedThreads;
+    private int primaryOrder;
+    private boolean timedout;
+    private int secondaryOrder;
+
+    /**
+     * Create a FlexibleBarrier and size it based on the number on contexts in `azt`.
+     */
+    FlexibleBarrier(AbstractZucchiniTest azt) {
+        this(azt, azt.contexts.size());
+    }
+
+    /**
+     * Create a FlexibleBarrier and size it based on `size` regardless of what is in `azt`.
+     *
+     * Future resizing is based on the contents of `azt`
+     */
+    FlexibleBarrier(AbstractZucchiniTest azt, int size) {
+        this.azt = azt;
+        this.primary = new Phaser(size);
+        this.secondary = new Phaser(size);
+        this.arrivedThreads = Collections.newSetFromMap(new ConcurrentHashMap<TestContext, Boolean>());
+        this.timedout = false;
+        this.primaryOrder = 0;
+        this.secondaryOrder = 0;
+    }
+
+    /**
+     * Kill all threads that have not reached this point and release all waiting threads.
+     */
+    void unlock() {
+        //force all late tests to fail
+        for(TestContext tc : this.azt.contexts) {
+            //if the thread has not arrived or already been registered as failed, register it as failed, and stop it
+
+            if(!this.arrivedThreads.contains(tc)) { //if it didn't successfully meet the barrier
+                if(!this.azt.failedContexts.contains(tc)) { //and it wasn't already registered as failed
+                    synchronized(azt.failedContexts) { //lock the contexts
+                        if(!this.azt.failedContexts.contains(tc)) { //test again for race condition prevention
+                            this.azt.failedContexts.add(tc);
+                            this.dec();
+                            if(tc.canKill) {
+                                tc.getThread().stop();
```

Comment:
It's unfortunately necessary, given that we are trying to prevent bad code from causing Zucchini to block on the barrier sync.  The Thread.interrupt() is only useful if the code within the thread is checking for interruptions or is performing IO operations.  

The Thread.stop() method will allow us to break from bad code such as 'while(true) {}'.  If I remove this, then we can't enforce the barrier timeout. @trentontrees